core/validation/specific_validations/towns_fund_round_four.py

```diff
@@ -268,6 +270,86 @@ def validate_leading_factor_of_delay(
+    fund_type = programme_id[:2]
+    project_ids = workbook["Project Details"]["Project ID"]
+    funding_df = workbook["Funding"]
+
+    # pull funding spent for individual projects and programme wide total
+    funding_spent = {project: spend_per_project(funding_df, project) for project in project_ids}
+    funding_spent[programme_id] = sum(funding_spent.values())
+
+    allocated_file = s3.get_file("data-store-file-assets-dev", f"{fund_type}-grant-awarded.csv")
+    funding_allocated = pd.read_csv(allocated_file, index_col="Index Code")["Grant Awarded"]
+
+    funding_spent_failures = []
+    if fund_type == "HS":
+        # check funding against programme wide funding allocated for Future High Street Fund submissions
+        ids_to_check = [programme_id]
+        for idx in ids_to_check:
+            if funding_spent[idx] > funding_allocated[idx]:
+                funding_spent_failures.append(
+                    TownsFundRoundFourValidationFailure(
+                        tab="Funding Profiles",
+                        section="Project Funding Profiles",
+                        message=(
+                            f"[Template message] You spent {funding_spent[idx]} "
+                            f"but were only allocated {funding_allocated[idx]}"
+                        ),
+                    )
+                )
+    else:
+        # check funding against individual project funding allocated for Towns Deal submissions
+        ids_to_check = project_ids
+        for idx in ids_to_check:
+            if funding_spent[idx] > funding_allocated[idx]:
+                funding_spent_failures.append(
+                    TownsFundRoundFourValidationFailure(
+                        tab="Funding Profiles",
+                        section=f"Project Funding Profiles - Project {get_project_number(idx)}",
+                        message=(
+                            f"[Template message] You spent {funding_spent[idx]} "
+                            f"but were only allocated {funding_allocated[idx]}"
+                        ),
+                    )
+                )
+
+    return funding_spent_failures
+
+
+def spend_per_project(funding_df: pd.DataFrame, project_id: str) -> float:
+    """return the total funding spent per an individual project
+
+    :param funding_df: A dataframe of the funding table from a submission
+    :param: project_id: ID of project in question
+
+    :return: Total funding spent per project
+    """
+    funding_spent = (
+        funding_df["Spend for Reporting Period"]
+        .loc[
+            (funding_df["Project ID"] == project_id)
+            & (funding_df["Funding Source Type"] == "Towns Fund")
+            & ~(funding_df["Funding Source Name"].str.contains(r"contractually committed"))
+        ]
+        .sum()
+    )
+    # Business logic here is taken from spreadsheet 4a - Funding Profile Z45 for grand total expenditure
+    return funding_spent
+
+
 @dataclass
```

Comment:
Does a function like this need to go in utils? It is only used onece and put here because it is a massive piec of logic. Could add some tests just for this although they are covered in the parent function tests

Reply:
I think it's too specific to go in utils so is fine here. I agree that the parent function tests should cover this fine as well